File: go/src/monkey/repl/repl.go
package repl

import (
	"bufio"
	"fmt"
	"io"
	"monkey/compiler"
	"monkey/evaluator"
	"monkey/lexer"
	"monkey/object"
	"monkey/parser"
	"monkey/vm"
)

const prompt = ">> "

// Start the loop of the REPL
func Start(in io.Reader, out io.Writer) {
	scanner := bufio.NewScanner(in)
	// env := object.NewEnvironment()
	macroEnv := object.NewEnvironment()

	constants := []object.Object{}
	globals := make([]object.Object, vm.GlobalsSize)

	symbolTable := compiler.NewSymbolTable()
	for i, v := range object.BuiltIns {
		symbolTable.DefineBuiltIn(i, v.Name)
	}

	for {
		fmt.Fprintf(out, prompt)
		scanned := scanner.Scan()

		if !scanned {
			return
		}

		line := scanner.Text()

		l := lexer.New(line)
		p := parser.New(l)

		program := p.ParseProgram()

		if len(p.Errors()) != 0 {
			printParserErrors(out, p.Errors())
			continue
		}

		evaluator.DefineMacros(program, macroEnv)
		expanded := evaluator.ExpandMacros(program, macroEnv)

		comp := compiler.NewWithState(symbolTable, constants)
		err := comp.Compile(expanded)
		if err != nil {
			io.WriteString(out, fmt.Sprintf("Whoops! Compilation failed:\n%s\n", err))
			continue
		}

		code := comp.Bytecode()
		constants = code.Constants

		machine := vm.NewWithGlobalsStore(code, globals)
		err = machine.Run()
		if err != nil {
			io.WriteString(out, fmt.Sprintf("Whoops! Executing bytecode failed:\n%s\n", err))
			continue
		}

		lastPopped := machine.LastPoppedStackElem()
		if lastPopped != nil {
			io.WriteString(out, lastPopped.Inspect())
			io.WriteString(out, "\n")
		}
	}
}

const monkeyFace = `            __,__
   .--.  .-"     "-.  .--.
  / .. \/  .-. .-.  \/ .. \
 | |  '|  /   Y   \  |'  | |
 | \   \  \ 0 | 0 /  /   / |
  \ '- ,\.-"""""""-./, -' /
   ''-' /_   ^ ^   _\ '-''
       |  \._   _./  |
       \   \ '~' /   /
        '._ '-=-' _.'
           '-----'
`

func printParserErrors(out io.Writer, errors []string) {
	io.WriteString(out, monkeyFace)
	io.WriteString(out, "Woops! We ran into some monkey business here!\n")
	io.WriteString(out, " parser errors:\n")

	for _, msg := range errors {
		io.WriteString(out, fmt.Sprintf("\t%s\n", msg))
	}
}

File: rust/rmonkey/src/repl.rs
use crate::object::Object;
use crate::vm::new_globals;
use std::io;
use std::io::BufRead;
use std::io::Write;

use crate::compiler::{Compiler, SymbolTable};
use crate::lexer::Lexer;
use crate::parser::Parser;
use crate::vm::VM;
use std::rc::Rc;

pub fn run() {
    let stdin = io::stdin();

    let mut constants: Vec<Rc<Object>> = vec![];
    let mut globals = new_globals();
    let mut symbol_table = SymbolTable::new_with_builtins();

    loop {
        print!(">> ");
        io::stdout().flush().expect("Error flushing stdout");

        let mut line = String::new();
        stdin
            .lock()
            .read_line(&mut line)
            .expect("Error reading from stdin");
        let lexer = Lexer::new(&line);

        let mut parser = Parser::new(lexer);

        let program = parser.parse_program();

        if !parser.errors().is_empty() {
            print_parser_errors(parser);
            continue;
        }

        let mut compiler = Compiler::new_with_state(symbol_table, constants);

        match compiler.compile(&program) {
            Ok(_) => {
                let mut vm = VM::new_with_globals_store(
                    &compiler.constants,
                    compiler.current_instructions().clone(),
                    globals,
                );

                match vm.run() {
                    Ok(result) => println!("{}", result),
                    Err(err) => println!("Whoops! Executing bytecode failed:\n{}\n", err),
                }

                globals = vm.globals;
            }
            Err(err) => println!("Whoops! Compilation failed:\n{}\n", err),
        };

        symbol_table = compiler.symbol_table;
        constants = compiler.constants;
    }
}

fn print_parser_errors(parser: Parser) {
    println!(
        r#"            __,__
.--.  .-"     "-.  .--.
/ .. \/  .-. .-.  \/ .. \
| |  '|  /   Y   \  |'  | |
| \   \  \ 0 | 0 /  /   / |
\ '- ,\.-"""""""-./, -' /
''-' /_   ^ ^   _\ '-''
   |  \._   _./  |
   \   \ '~' /   /
    '._ '-=-' _.'
       '-----'
"#
    );
    println!("Woops! We ran into some monkey business here!");
    for error in parser.errors() {
        println!("\t{:?}", error);
    }
}
